### PR TITLE
Detect invalid RangeAttribute arguments. Fixes #472

### LINKED ISSUE #472: Overflow exception and DivideByZero exception from the RangeAttribute
Specific data values fed to the RangeAttribute can trigger a DivideByZero exception and/or an OverflowException. These values have been pointed out by the Smart Unit Tests feature from Visual Studio 2015, and an image with the data values is attached.

The DivideByZero may happen by the user entering 0 as a step value.
The OverflowException is more serious as it highlights an implicit constraint saying “to > from” that is not stated anywhere. There should there be any need for that.
exceptions.

These issues are seen for all the RangeAttribute overloads.
Please let me know if you need any further information.

Pratap Lakshman

### PR DESCRIPTION
**Pull request by CharliePoole**
This was laying around on my hard drive so I decided to clean it up and get it in.

I implemented the three points listed in issue #472:

1. Zero step gives an ArgumentException

2. Negative step with from < to or positive with from > to gives an Argument exception.

3. Other stuff works leaving aside Terje's comment about too small floating steps causing overflow, which is not handled in the current code.

It would be very easy to make bullet 2 work as well, by simply reversing the sign of the step. However, I haven't done that.

Docs need updating.

### CODE REVIEW

File: src/NUnitFramework/framework/Attributes/RangeAttribute.cs [hunk 1]
```diff
@@ -47,6 +47,9 @@ public class RangeAttribute : ValuesAttribute
         /// <param name="step"></param>
         public RangeAttribute(int from, int to, int step)
         {
+            Guard.ArgumentValid(step > 0 && to >= from || step < 0 && to <= from,
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/RangeAttribute.cs]: Does it matter what the step is if `from == to`?

**Comment by rprouse** [comment]
Everything looks good except for my one comment which applies to all the guard statements.

File: src/NUnitFramework/framework/Attributes/RangeAttribute.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Attributes/RangeAttribute.cs]: It doesn't.

Comment (CharliePoole) [inline src/NUnitFramework/framework/Attributes/RangeAttribute.cs]: Not sure if you're suggesting a change here.

Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/RangeAttribute.cs]: Me bad, sorry, I am so used to if(...) throw that I inverted the guard in my head. This is right, I am merging.

**Pull request merged**